Processes for WCL based on RFC 330 (#2905)
sourcegraph/about: handbook/engineering/web/frontend-platform/wildcard-component-library.md

# Wildcard Component Library (WCL)

Wildcard Component Library is a comprehensive collection of reusable, design-approved components that can be composed to quickly build out different user journeys across the web app. WCL is stored in its own [directory](https://github.com/sourcegraph/sourcegraph/tree/main/client/wildcard), owned by the [Fronted Platform Team](https://about.sourcegraph.com/handbook/engineering/web/frontend-platform) and contributed to by all Sourcegraph teammates. WCL is a part of the Wildcard Design System.

## WCL x Frontend Platform Team
- The Frontend Platform Team is the owner of the design system and Wildcard Component Library  and needs to be up-to-date with all the changes and needs of other teams
- The Frontend Platform Team should enable other teams to make changes to the Wildcard Component Library by keeping them up-to-date (see Sharing updates to the components system), helping them plan their work, providing expertise, and reviewing the changes

> - The Frontend Platform team **shouldn't be responsible for implementing components for other teams**. Otherwise, we would create a bottleneck and FE Platform team wouldn’t have time to focus on their team goals.

Reviewer: ```suggestion
- The Frontend Platform team **shouldn't be exclusively responsible for implementing components for other teams**. Otherwise, we would create a bottleneck and the FE Platform team wouldn’t have time to focus on their team goals.
```

I think it is OK for the team to sometimes build components for other teams if it aligns with our roadmap or it introduces a new behavior/pattern that would be better if it was started by the FPT.

Reviewer 2: I agree with @jeanduplessis's suggestion. The frontend platform team will allocate time for adding components and documentation to the Wildcard library, but it should not be solely responsible for doing that. 

> This is a serious risk too for our design/UI quality because this means over time the UI becomes inconsistent

Improving the Wildcard library docs and processes around introducing new components should help with that. FTP should facility best practices and make product teams aware of them. Another direction that we should look at is making our linters sticker to ensure that there's one correct way of doing things. Consistency shouldn't require the full attention of some particular team, but rather be the result of the processes and practices established for the front-end development.

> - The Frontend Platform team **shouldn't be exclusively responsible for implementing components for other teams**. Otherwise, we would create a bottleneck and the FE Platform team wouldn’t have time to focus on their team goals.

Reviewer 3: Here's my suggestion:
```suggestion
- The Frontend Platform team isn't **exclusively** responsible for implementing components for other teams. Otherwise, we would create a bottleneck and the FE Platform team wouldn’t have time to focus on their other team goals.
```

Author: @pdubroy wdyt about it?

Reviewer 4: lgtm. I will also add a suggestion for the next bullet point that adds more detail...will just take a couple mins.

> - Other product teams should feel empowered to make changes and create components by themselves

Reviewer 4: ```suggestion
- Other product teams should feel empowered to make changes and create components by themselves, with the Frontend Platform team in a consulting (and ultimately, approving) role.
```

## Changing and creating components
What steps to take when one of the teams needs to update or add a component in WCL?
1. The designer creates a Figma file with full specification regarding the component (separate from the general project file or general design system files) and a Github issue for a new component to track the implementation progress
2. The designer initiates the cross-team communication by marking as reviewers:

> - Members of the engineering team who will be implementing the work
> - At least one other Design Team member
> - Members of other teams affected by the changes

Reviewer 3: ```suggestion
   - Members of the engineering team who will be implementing the work
   - At least one other Design Team member
   - Members of other teams affected by the changes
```

3. After all the reviews are completed, the engineering team plans the implementation. The engineering lead should create a proposal for a new component on the [FE Platform Project board](https://github.com/orgs/sourcegraph/projects/144) using [Wildcard proposal template](https://github.com/sourcegraph/sourcegraph/blob/main/.github/ISSUE_TEMPLATE/wildcard_proposal.md).
4. During the implementation, the engineering team stays in constant communication with the Design Team members engaged in the project. Asks for review, if necessary.
5. The engineering team should ask for final reviews from:

> - Members of the FE Platform team
> - Members of the engineering team implementing the change
> - Members of other teams affected by the changes (if applicable)
> - Members of the Design Team engaged in the project

Reviewer 3: Looking at the rendered view, this wasn't indented correctly
```suggestion
   - Members of the FE Platform team
   - Members of the engineering team implementing the change
   - Members of other teams affected by the changes (if applicable)
   - Members of the Design Team engaged in the project
```

# Wildcard Component Library (WCL)

> Wildcard Component Library is a comprehensive collection of reusable, design-approved components that can be composed to quickly build out different user journeys across the web app. WCL is stored in its own [directory](https://github.com/sourcegraph/sourcegraph/tree/main/client/wildcard), owned by the [Fronted Platform Team](https://about.sourcegraph.com/handbook/engineering/web/frontend-platform) and contributed to by all Sourcegraphers. WCL is a part of the Wildcard Design System.

Reviewer: ```suggestion
Wildcard Component Library is a comprehensive collection of reusable, design-approved components that can be composed to quickly build out different user journeys across the web app. WCL is stored in its own [directory](https://github.com/sourcegraph/sourcegraph/tree/main/client/wildcard), owned by the [Fronted Platform Team](https://about.sourcegraph.com/handbook/engineering/web/frontend-platform) and contributed to by all Sourcegraph teammates. WCL is a part of the Wildcard Design System.
```

> Wildcard Component Library is a comprehensive collection of reusable, design-approved components that can be composed to quickly build out different user journeys across the web app. WCL is stored in its own [directory](https://github.com/sourcegraph/sourcegraph/tree/main/client/wildcard), owned by the [Fronted Platform Team](https://about.sourcegraph.com/handbook/engineering/web/frontend-platform) and contributed to by all Sourcegraphers. WCL is a part of the Wildcard Design System.

Reviewer 2: > across the web app

Should we indicate here that we might want to use these reusable components outside of the web application?

Author: @valerybugakov could you suggest the change? Thanks!

Reviewer 5: How about `...that can be composed to quickly build out different user journeys across our web applications.`?

## WCL x Frontend Platfrom Team

> - The Frontend Platform Team is the owner of the design system and Wildcard Component Library components and needs to be up-to-date with all the changes and needs of other teams

Reviewer: ```suggestion
- The Frontend Platform Team is the owner of the design system and Wildcard Component Library  and needs to be up-to-date with all the changes and needs of other teams
```

> - The Frontend Platform Team should enable other teams make changes to the Wildcard Component Library by keeping them up-to-date (see Sharing updates to the components system), helping them plan their work, providing expertise, and reviewing the changes

Reviewer: ```suggestion
- The Frontend Platform Team should enable other teams to make changes to the Wildcard Component Library by keeping them up-to-date (see Sharing updates to the components system), helping them plan their work, providing expertise, and reviewing the changes
```

> - The Frontend Platform Team should enable other teams make changes to the Wildcard Component Library by keeping them up-to-date (see Sharing updates to the components system), helping them plan their work, providing expertise, and reviewing the changes

Reviewer 2: Not a native speaker here, but it seems right: 

```suggestion
- The Frontend Platform Team should enable other teams to make changes to the Wildcard Component Library by keeping them up-to-date (see Sharing updates to the components system), helping them plan their work, providing expertise, and reviewing the changes
```

> - The Frontend Platform Team should enable other teams make changes to the Wildcard Component Library by keeping them up-to-date (see Sharing updates to the components system), helping them plan their work, providing expertise, and reviewing the changes

Author: @pdubroy the link you've pasted links to this issue. Could you please update it? Thank you!

> - The Fronted Platform Team **shouldn’t be responsible for implementing components for other teams**. Otherwise, we would create a bottleneck and FE Platform Team wouldn’t have time to focus on their team goals.

Reviewer: ```suggestion
- The Frontend Platform team **shouldn't be responsible for implementing components for other teams**. Otherwise, we would create a bottleneck and FE Platform team wouldn’t have time to focus on their team goals.
```

> - The Fronted Platform Team **shouldn’t be responsible for implementing components for other teams**. Otherwise, we would create a bottleneck and FE Platform Team wouldn’t have time to focus on their team goals.

Reviewer 2: Should we put `.` at the end of each sentence then?

- Other product teams should feel empowered to make changes and create components by themselves

## Changing and creating components

> ### Process of changing or creating a component by a product team

Reviewer 5: I can't remember if I've discussed this before. It feels a bit off that we have two separate processes depending on where the change originates from. The processes are quite similar, and I feel it opens up a grey area for work that is a collaboration across teams.

### Process of changing or creating a component

1. The designer creates a Figma file with full specification regarding the component (separate from the general project file or general design system files)
2. The designer initiates the cross-team communication by marking as reviewers:
  - Members of the engineering team who will be implementing the work
  - At least one other Design Team member
  - Members of other teams affected by the changes 
3. After all the reviews are completed, the engineering team plans the implementation
4. During the implementation, the engineering team stays in constant communication with the Design Team members engaged in the project. Asks for review, if necessary.
5. The engineering team should ask for final reviews from:
  - Members of the FE Platform team
  - Members of the engineering team implementing the change
  - Members of other teams affected by the changes (if applicable)
  - Members of the Design Team engaged in the project
7. When changes are approved, the engineering team merges the PR

Reviewer 2: As discussed at the Frontend Crew meeting, we also want to create a Github issue for a new component to track the implementation progress.

> What steps to take when one of the teams needs to update or add a Standard UI Component Library?

Reviewer: ```suggestion
What steps to take when one of the teams needs to update or add a component in WCL?
```

1. The lead designer for the project creates a separate page in the project Figma file with full specification regarding the component (separate from the general project page or general design system files)
2. The designer initiates the communication between the engineers from the product team and the Frontend Platform Team by marking them as reviewers in the Figma file containing the component specs

> 3. Engineers from both teams review the document and plan the implementation work. Both teams can choose engineering leads for the project who will be responsible for communication and progress (higher efficiency and clear ownership).

Reviewer 2: It would be great to mention here an issue template that @umpox added recently. Something like:

- Engineering lead should create an a proposal for a new component on the [FE Platform Project board](https://github.com/orgs/sourcegraph/projects/144) using [Wildcard proposal template](https://github.com/sourcegraph/sourcegraph/blob/main/.github/ISSUE_TEMPLATE/wildcard_proposal.md).

4. During the implementation, the FE Platform Team is informed about key milestones and any issues that occur. FE Platform engineers should be asked for a review in any of those key moments.
5. When the component is ready, the product team should ask for additional reviews from:

> - The lead engineer for this project from the FE Platform Team
> - The FE Platform Team Designer

Reviewer: ```suggestion
  - The lead engineer for this project from the FE Platform Team
  - The FE Platform Team Designer
```

6. When changes are approved, the product team merges the PR

Important: we would like to avoid the situation where the FE Platform Team is engaged only at the end of the process or doesn’t know about the changes at all.

### Process of changing or creating a component by the FE Platform Team or the Design Team

> It is possible that the Design Team or the FE Platform Team themselves initiate changes to the Standard UI Component Library. Here are the steps to take:

Reviewer: ```suggestion
It is possible that the Design Team or the FE Platform Team themselves initiate changes to the WCL. Here are the steps to take:
```

1. The designer creates a Figma file with full specification regarding the component (separate from the general project file or general design system files)
2. The designer initiates the cross-team communication by marking as reviewers:

> - The FE Platform Team engineers
> - At least one other Design Team member
> - Members of other teams affected by the changes 

Reviewer: ```suggestion
  - The FE Platform Team engineers
  - At least one other Design Team member
  - Members of other teams affected by the changes 
```

3. After all the reviews are completed, the FE Platform Team plans the implementation
4. During the implementation, the FE Platform Team stays in constant communication with the Design Team members engaged in the project. Asks for review, if necessary.
5. The FE Platform Team should ask for final reviews from:

> - The FE Platform Team members
> - The Design Team members engaged in the project

Reviewer: ```suggestion
  - The FE Platform Team members
  - The Design Team members engaged in the project
```

6. Members of other teams engaged by the project and affected by the changes 
7. When changes are approved, the FE Platform Team team merges the PR

> If product team after having started working on a feature, identifies in code review that a part of the UI should be extracted into a shared component (e.g., "This is very similar to the widget used in XYZ...can we make it a shared component?"), do not block the PR and kick off the "changing or creating a component" flow asynchronously.

Reviewer: ```suggestion
If the product team after having started working on a feature, identifies in a code review that a part of the UI should be extracted into a shared component (e.g., "This is very similar to the widget used in XYZ...can we make it a shared component?"), do not block the PR and kick off the "changing or creating a component" flow asynchronously.
```

## Definitions
### Full design specification should include:
- All variants of the component (ex. Filled/empty or for auth/unauth user)
- Clear descriptions of component behaviour added as notes on Figma
- Clear descriptions of specific changes added as notes on Figma. This is primarily for changes to a design which might not be immediately obvious to a viewer, like a small change in the positioning of a button, a wording change etc
- Designs or descriptions for different component states (e.g. hover/focus/animations). If it’s not possible to accurately represent an animation on Figma we should add a note to the design file explaining it (or linking to an example).
- Light/Dark mode designs (specially important when changing colours)
- Mobile/Desktop designs

### A component is deemed to be ready from an engineering perspective when effort has been made to ensure it is:
- Unit tested

> - Accessible

Reviewer: We should link to our accessibility standards here

- Documented textually, explaining when this component should be used and why
- In Storybook
- Built with a clear case for reusability

## Sharing updates to the components system
To keep other teams informed about the changes to the component system, we should share updates:

> - In the FE Platform Team weekly email

Reviewer: This is no longer accurate

Author: How about:
- In the FE Platform team [status update](https://about.sourcegraph.com/handbook/engineering/leadership#status-updates)

Reviewer: LGTM, updated my comment with a suggestion to apply.

> - In the FE Platform Team weekly email

Reviewer 2: Do we already have it? 😃

- In the #progress channel
- In a standalone documentation for the Wildcard Component Library (TODO)
- In the #components-system-updates channel

> - The component system is versioned, and changelogs are generated accordingly

Reviewer 2: Should we add `(TODO)` here too? 
It's not implemented yet.

Reviewer 5: I'd say to keep this consistent, also changing it to:

```suggestion
- Through generated changelogs (TODO)
```

I think that will be the closest we will get to typical versioning unless we decide to publish this library to NPM (not in scope right now)
